tinygrad/renderer/wgsl.py

```diff
@@ -0,0 +1,93 @@
+  (UPat(Ops.LOAD, name="root", dtype=dtypes.bool, src=(UPat())), lambda root: UOp(root.op, dtypes.int, root.src, root.arg).cast(dtypes.bool)),
+  (UPat(GroupOp.ALU, src=(UPat(name="a"), UPat(name="b")), name="c"), fixup_binops),
+  *[(UPat(a, src=(UPat(name="b", dtype=(dtypes.uint, dtypes.int, dtypes.bool))), name="a"),
+     lambda a,b: UOp(a, dtypes.float, (b.cast(dtypes.float),)).cast(b.dtype))
+    for a in (UnaryOps.EXP2, UnaryOps.SIN, UnaryOps.LOG2, UnaryOps.SQRT)],
+  (UPat.store(UPat.var("bidx"), UPat.var("var", dtype=dtypes.bool), UPat.var("gate")),
+   lambda bidx,val,gate: UOp.store(bidx, val.cast(dtypes.int), gate)),
+  (UPat.store(UPat.var("bidx"), UPat.var("var", dtype=dtypes.bool)), lambda bidx,var: UOp.store(bidx, var.cast(dtypes.int))),
+  (UPat(Ops.MAX, name="m"), lambda m: (m.src[0] < m.src[1]).where(m.src[1], m.src[0])),
+  # fix nan propagation: 'a * select(1, nan, cond) -> select(a, nan, cond)'
+  (UPat(Ops.MUL, name="m", src=(UPat(name="a"), UPat(TernaryOps.WHERE, src=(UPat.var("g"), \
+    UPat(op=Ops.CONST, name="c1"), UPat(op=Ops.CONST, name="c2"))))), \
+    lambda m,a,g,c1,c2: UOp(TernaryOps.WHERE, dtype=m.dtype, src=(g, UOp.const(dtype=dtypes.float, b=float('nan')), a)) \
+    if math.isnan(c1.arg) and c2.arg == 1.0 else None),
+  ])
+
+type_map = { dtypes.float: "f32", dtypes.uchar: "u32", dtypes.ushort: "u32", dtypes.short: "i32",
+            dtypes.char: "i32", dtypes.int32: "i32", dtypes.uint32: "u32", dtypes.bool: "bool" }
+
+# convert from pointer style indexing to array style
+def render_load_store(r, bidx, sext = False, sext_am = 8):
+  sbidx = strip_parens(r[bidx])
+  buf, idx = sbidx.split("+")[0], '+'.join(sbidx.split("+")[1:])
+  # sign-extend when loading char/short
+  return f"bitcast<i32>(select(0u, 0xffffffffu << {sext_am}, (({buf}[{idx}] >> {sext_am-1}) > 0)) | bitcast<u32>({buf}[{idx}]))" \
+    if sext else f"{buf}[{idx}]"
+
+class WGSLRenderer(CStyleLanguage):
+  device = "WEBGPU"
+  global_max = (65535, 65535, 65535)
+  local_max = (256, 256, 64)
+  code_for_workitem = {"g": lambda x: f"i32(gindex.{'xyz'[int(x)]})", "l": lambda x: f"i32(lindex.{'xyz'[int(x)]})"}
+  extra_matcher = wgsl_matcher
+  external_local_bufs = True
+  supports_float4 = False
+  barrier = "workgroupBarrier();"
+  code_for_op = {**CStyleLanguage.code_for_op, TernaryOps.WHERE: lambda a,b,c,dtype: f"select({c},{b},{a})"}
+  nan = "nan()"
+  type_map = type_map
+
+  string_rewrite = PatternMatcher([
+    (UPat(Ops.CONST, dtype=dtypes.bool, name="x"), lambda ctx,x: "true" if x.arg else "false"),
+    (UPat(Ops.CONST, dtype=(dtypes.char, dtypes.short), name="x"), lambda ctx,x: f"i32({x.arg})"),
+    (UPat(Ops.CONST, dtype=(dtypes.uchar, dtypes.ushort, dtypes.uint32), name="x"), lambda ctx,x: f"bitcast<u32>({x.arg}i)" \
+     if x.arg < 0 else f"{x.arg&0xFFFFFFFF}u"),
+    (UPat(Ops.CONST, arg=math.inf, name="x"), lambda ctx, x: f"{type_map[x.dtype]}({ctx.infinity})"),
```

Comment:
This shouldn't be duplicated because of the type thing

Reply:
I propose a `render_cast` in cstyle: use it in the `CONST` string rewrites where we cast, and override `render_cast` in wgsl. This adds 1 line in cstyle, 1 line in wgsl, and removes 4 from wgsl, so -2 overall.

Reply:
https://github.com/tinygrad/tinygrad/pull/7687